fastlane_bot/helpers/wrap_unwrap_processor.py

```diff
@@ -0,0 +1,344 @@
+
+        This function checks if there is a leftover balance of native/wrapped gas tokens at the end of the trade actions, and adds a final wrap/unwrap trade to convert the balance into the flashloan token.
+
+        Args:
+            deadline: the deadline for the trade to be completed by.
+            new_route_struct: the processed list of trade actions, as dictionaries.
+
+
+        """
+        should_wrap = (
+            self.flashloan_wrapped_gas_token
+            and self.balance_tracker.get(self.cfg.NATIVE_GAS_TOKEN_ADDRESS, 0) > 0
+        )
+        should_unwrap = (
+            self.flashloan_native_gas_token
+            and self.balance_tracker.get(self.cfg.WRAPPED_GAS_TOKEN_ADDRESS, 0) > 0
+        )
+
+        if should_wrap or should_unwrap:
+            new_route_struct.append(
+                self._get_wrap_or_unwrap_native_gas_tkn_struct(
+                    deadline=deadline, wrap=should_wrap or not should_unwrap
+                )
+            )
+
+    def _get_wrap_or_unwrap_native_gas_tkn_struct(
+        self, deadline: int, wrap: bool, source_amount: int = 0
+    ):
+        """
+        This function provides a trade struct to wrap or unwrap the native gas token.
+
+        :param deadline: the deadline by which the trade must be executed.
+        :param wrap: if True, wrap the native gas token, else unwrap the native gas token.
+        :param source_amount: the number of tokens to wrap or unwrap. If 0, this will wrap or unwrap the current balance of tokens.
+
+        returns: A dictionary containing the wrap or unwrap trade instructions that will be sent to the contract.
+        For example:
+            {
+            "platformId": 10,
+            "targetToken": "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
+            "sourceToken": "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",
+            "sourceAmount": "1000000000000000000",
+            "minTargetAmount": "1000000000000000000",
+            "deadline": "124124124",
+            "customAddress": "0x0000000000000000000000000000000000000000",
+            "customInt": 0,
+            "customData": "0x",
+            }
+
+        """
+        return asdict(
+            RouteStruct(
+                platformId=self.cfg.PLATFORM_ID_WRAP_UNWRAP,
+                targetToken=self.cfg.WRAPPED_GAS_TOKEN_ADDRESS
+                if wrap
+                else self.cfg.NATIVE_GAS_TOKEN_ADDRESS,
+                sourceToken=self.cfg.NATIVE_GAS_TOKEN_ADDRESS
+                if wrap
+                else self.cfg.WRAPPED_GAS_TOKEN_ADDRESS,
+                sourceAmount=int(source_amount),
+                minTargetAmount=int(source_amount),
+                deadline=deadline,
+                customAddress=self.cfg.WRAPPED_GAS_TOKEN_ADDRESS
+                if wrap
+                else self.cfg.NATIVE_GAS_TOKEN_ADDRESS,
+                customInt=0,
+                customData="0x",
+            )
+        )
+
+    class FlashloanTokenException(Exception):
```

Comment:
If this is a new exception class, we should get in the habit of defining these in a dedicated `exceptions.py` module. If it's not newly defined in this PR you can disregard.

Reply:
Moved FlashloanTokenException -> exceptions.py